### ISSUE TITLE
Dapr lifecycle management improve - running time

### ISSUE THREAD

**Issue by cshi0** [dapr/dapr#3376]
## In what area(s)?

/area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release
## Describe the proposal

This proposal is a complement to #2918 that defines the behaviors of dapr when the application is down and restarts at running time.

**Comment by daixiang0** [comment]
Cool, really happy to see it accepted!

What I am curious is when app failed, how dapr handle input/output data? Cache them or leave it at binding side?

**Comment by cshi0** [comment]
Dapr leaves it at binding side, and dapr does not promise anything. Some components retry after failure, and it is why a short grace time is put before unsubscribe inbound traffic. However, it is not promised, and messages might loss because of the app crash and specific binding side design.

**Comment by yaron2** [comment]
@cshi0 what's the reasoning behind this? to allow other consumers in a possible consumer group to get messages delivered to them immediately?

**Comment by cshi0** [comment]
@yaron2 Yes. It is the biggest reason.

**Comment by skyao** [comment]
> @cshi0 what's the reasoning behind this? to allow other consumers in a possible consumer group to get messages delivered to them immediately?

Yes, yaron, you get the point. This is very important for pub/sub, if one instance is down, the other instances in the same consumer group  should continue to consumer.

**Comment by withinboredom** [comment]
Can you explain more about why this is necessary?

- Kubernetes will kill the container if the liveness probe fails. By not having a restart policy, `daprd` will also be killed when the pod is killed due to the failed container.
- If the application fails, it should die with a non-zero exit and kubernetes will automatically schedule a new pod and kill `daprd` container (via killing the pod).

Is this for working with containers in a pod that have a restart policy?

**Comment by cshi0** [comment]
@withinboredom 
> Is this for working with containers in a pod that have a restart policy?
Yes. Besides that, an application might become unavailable even it is alive to the container, and in that case, dapr should not give it any requests.

**Comment by withinboredom** [comment]
Could this then exacerbate a misconfiguration? Say an app can only accept 100 concurrent requests but `daprd` isn't configured correctly. Currently, IIRC, in this case, Dapr would timeout or wait forever to try the 101st request, eventually the app would make progress and accept it.

If this proposal goes through, a 100 concurrent request app would now have to account for the probes and normal work. (This is already the case when using liveness probes).

This makes me realize that instead of using probes, maybe we can just assign a timeout to connections to the app. If configured, apps that have a string of "failed connections" would be considered "dead" and use this flow.

**Comment by cshi0** [comment]
@withinboredom Yes, we can use a timeout to determine a failed connection. However, I want to recover the connection between daprd and inbound traffic sources like kafka or rocketmq after the app is alive again. In this case, I think I have to continuously probe the status of the app (or the container).

**Comment by cshi0** [comment]
The design of probing process is slightly changed as I mentioned in the PR.

**Comment by artursouza** [comment]
> When the application is temporarily unavailable, dapr needs to unsubscribe inbound traffic (subscription and input binding) to avoid the loss of messages because the down time of application might be longer than the retry timeout for sending messages. Unsubscribing inbound traffic can also make the service available because all the following messages will be processed by running applications.
> 
> Dapr will also need the ability to detect the availability of the application so that it can resubscribe inbound traffic when the application is available again.


This seems to be referring to circuit breaking. @pkedy is looking into that space as well.

**Comment by cshi0** [comment]
An implementation of the proposal is submitted as a PR #3403 .

**Comment by artursouza** [comment]
Can you show an example of loss of message? Dapr guarantees at least once delivery. Also @pkedy is working on circuit breaking feature in Dapr, which overlaps with this issue.
